src/compiler/binder.ts

```diff
@@ -1542,22 +1589,79 @@ namespace ts {
             }
         }
 
-        function bindCallExpressionFlow(node: CallExpression) {
-            // If the target of the call expression is a function expression or arrow function we have
-            // an immediately invoked function expression (IIFE). Initialize the flowNode property to
-            // the current control flow (which includes evaluation of the IIFE arguments).
-            let expr: Expression = node.expression;
-            while (expr.kind === SyntaxKind.ParenthesizedExpression) {
-                expr = (<ParenthesizedExpression>expr).expression;
-            }
-            if (expr.kind === SyntaxKind.FunctionExpression || expr.kind === SyntaxKind.ArrowFunction) {
-                bindEach(node.typeArguments);
-                bindEach(node.arguments);
-                bind(node.expression);
+        function bindOptionalExpression(node: Expression, presentTarget: FlowLabel, missingTarget: FlowLabel) {
+            const savedPresentTarget = currentPresentTarget;
+            const savedMissingTarget = currentMissingTarget;
+            currentPresentTarget = presentTarget;
+            currentMissingTarget = missingTarget;
+            bind(node);
+            currentPresentTarget = savedPresentTarget;
+            currentMissingTarget = savedMissingTarget;
+            if (!isValidOptionalChain(node) || isOutermostOptionalChain(node)) {
+                addAntecedent(presentTarget, createFlowOptionalChain(FlowFlags.Present, currentFlow, node));
+                addAntecedent(missingTarget, createFlowOptionalChain(FlowFlags.Missing, currentFlow, node));
+            }
+        }
+
+        function isOutermostOptionalChain(node: ValidOptionalChain) {
+            return !isOptionalChain(node.parent) || !!node.parent.questionDotToken || node.parent.expression !== node;

```

Comment:
Does this definition of outermost handle parenthesis? eg, `(((a.b)?.c)?.["d"])?.()` (do optional chains support that? they do, right?)

Reply:
Optional chains stop at parentheses (i.e. `(a?.b).c` could throw if `a` is `undefined`). Each optional chain in the CFA graph gets a Present/Missing branch from the `?.` until the end of the chain.